lib/task_005_lib.rb

```diff
@@ -0,0 +1,95 @@
+module Task005
+  module_function
+
+  class Point
+    attr_reader :x, :y
+
+    def self.from_input(point)
+      x, y = point.split(',').map(&:to_i)
+      Point.new(x, y)
+    end
+
+    def initialize(x, y)
+      @x = x
+      @y = y
+    end
+  end
+
+  class Line
+    attr_reader :point1, :point2
+
+    def self.from_input(input)
+      point1, point2 = input.split(' -> ').map { |point| Point.from_input(point) }
+      Line.new(point1, point2)
+    end
+
+    def initialize(point1, point2)
+      @point1 = point1
+      @point2 = point2
+    end
+
+    def horizontal_or_vertical?
+      @point1.x == @point2.x || @point1.y == @point2.y
+    end
+
+    def degree_45?
+      (@point1.x - @point2.x).abs == (@point1.y - @point2.y).abs
+    end
+
+    def points_on_line
+      if horizontal_or_vertical?
+        points = []
+        x1, x2 = [@point1.x, @point2.x].sort
+        y1, y2 = [@point1.y, @point2.y].sort
+        (x1..x2).each do |x|
+          (y1..y2).each do |y|
+            points << [x, y]
+          end
+        end
+      elsif degree_45?
+        points = []
+        x1, x2 = [@point1.x, @point2.x].sort
+        y1, y2 = [@point1.y, @point2.y].sort
+
+        x = @point1.x - @point2.x < 0 ? (x1..x2).each.to_a : (x1..x2).each.to_a.reverse
+        y = @point1.y - @point2.y < 0 ? (y1..y2).each.to_a : (y1..y2).each.to_a.reverse
+
+        x.size.times do |i|
+          points << [x[i], y[i]]
+        end
+      end
```

Comment:
You need to support 2 cases:
* horizontal/vertical lines,
* degree 45 lines.

As you probably know, there is a hidden third case: all the other lines. At the moment you returns `[]` which is not the real list of points on line. This is fine because you do not execute this code for such third case, so it should not be a problem. And my understanding is that running this `points_on_line` method for this third case would be considered by you as a bug / error in your code, right?

In such case, it would be better to raise a proper error. In the other words, fail loudly instead of returning `[]` and silently let the script keep working.

How can you raise an error? There is a `raise` method in ruby: https://ruby-doc.org/core-3.1.0/Kernel.html#method-i-raise.

Usually, you want to pass 2 arguments to `raise`: an error and a message.

For example:

```suggestion
      else
        raise NotImplementedError, "The lines other than horizontal/vertical/45 degree are not supported yet!"
      end
```

Reply:
I was thinking about kind of this, this is good to know about